ts/editor/ColorButtons.svelte

```diff
@@ -0,0 +1,50 @@
+<!--
+Copyright: Ankitects Pty Ltd and contributors
+License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+-->
+<script lang="typescript">
+    import * as tr from "lib/i18n";
+
+    import IconButton from "components/IconButton.svelte";
+    import ColorPicker from "components/ColorPicker.svelte";
+    import ButtonGroup from "components/ButtonGroup.svelte";
+    import WithShortcut from "components/WithShortcut.svelte";
+
+    import { squareFillIcon } from "./icons";
+
+    import "./color.css";
+
+    const foregroundColorKeyword = "--foreground-color";
+    let color = "black";
+
+    $: {
+        document.documentElement.style.setProperty(foregroundColorKeyword, color);
+    }
+
+    function wrapWithForecolor(): void {
+        document.execCommand("forecolor", false, color);
+    }
+
+    function setWithCurrentColor({ currentTarget }: Event): void {
+        color = (currentTarget as HTMLInputElement).value;
+    }
+</script>
+
+<ButtonGroup id="color">
+    <WithShortcut shortcut="F7" let:createShortcut let:shortcutLabel>
+        <IconButton
+            class="forecolor"
+            tooltip={`${tr.editingSetForegroundColor} (${shortcutLabel})`}
+            on:click={wrapWithForecolor}
```

Comment:
I like how this makes the wrapper's functionality more explicit, and we could potentially make the tooltip less cumbersome by returning a function that appends the label instead, eg:

```html
    <WithShortcut shortcut="F7" let:createShortcut let:withShortcutLabel>
        <IconButton
            tooltip={withShortcutLabel(tr.editingSetForegroundColor()}
```

That said, would it make sense to teach each button component to take a shortcut prop directly instead? That allows us to push more of the complexity into the components, and simplify the code that consumes them. Eg something like:

```diff
diff --git a/ts/components/IconButton.svelte b/ts/components/IconButton.svelte
index 56269538e..ab417beb6 100644
--- a/ts/components/IconButton.svelte
+++ b/ts/components/IconButton.svelte
@@ -4,19 +4,19 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 -->
 <script lang="typescript">
     import type { Readable } from "svelte/store";
-    import { hasContext, getContext, onMount, createEventDispatcher } from "svelte";
+    import { getContext, onMount } from "svelte";
     import { disabledKey, nightModeKey, dropdownKey } from "./contextKeys";
+    import { useShortcut } from "lib/shortcuts";
 
     export let id: string;
     let className = "";
     export { className as class };
 
     export let tooltip: string | undefined;
+    export let shortcut = "";
     export let active = false;
     export let disables = true;
 
-    let buttonRef: HTMLButtonElement;
-
     const disabled = getContext<Readable<boolean>>(disabledKey);
     $: _disabled = disables && $disabled;
 
@@ -25,8 +25,14 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
     const dropdown = getContext(dropdownKey);
     const dropdownProps = dropdown?.getDropdownTriggerProps() ?? {};
 
-    const dispatch = createEventDispatcher();
-    onMount(() => dispatch("mount", { button: buttonRef }));
+    let buttonRef: HTMLButtonElement;
+    let withShortcutSuffix: (string) => string;
+    $: tooltipWithShortcut = withShortcutSuffix?.(tooltip);
+    onMount(() => {
+        const handle = useShortcut(buttonRef, shortcut);
+        withShortcutSuffix = handle.withShortcutSuffix;
+        return handle.onDestroy;
+    });
 </script>
 
 <style lang="scss">
@@ -69,7 +75,7 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
     class:btn-day={!nightMode}
     class:btn-night={nightMode}
     tabindex="-1"
-    title={tooltip}
+    title={tooltipWithShortcut}
     disabled={_disabled}
     class:dropdown-toggle={Boolean(dropdown)}
     {...dropdownProps}
diff --git a/ts/editor/ColorButtons.svelte b/ts/editor/ColorButtons.svelte
index db3c2afd0..cbe2dc0a6 100644
--- a/ts/editor/ColorButtons.svelte
+++ b/ts/editor/ColorButtons.svelte
@@ -31,15 +31,13 @@ License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
 </script>
 
 <ButtonGroup id="color">
-    <WithShortcut shortcut="F7" let:createShortcut let:shortcutLabel>
-        <IconButton
-            class="forecolor"
-            tooltip={`${tr.editingSetForegroundColor} (${shortcutLabel})`}
-            on:click={wrapWithForecolor}
-            on:mount={createShortcut}>
-            {@html squareFillIcon}
-        </IconButton>
-    </WithShortcut>
+    <IconButton
+        class="forecolor"
+        tooltip={tr.editingSetForegroundColor()}
+        shortcut="F7"
+        on:click={wrapWithForecolor}>
+        {@html squareFillIcon}
+    </IconButton>
 
     <WithShortcut shortcut="F8" let:createShortcut let:shortcutLabel>
         <ColorPicker
diff --git a/ts/lib/shortcuts.ts b/ts/lib/shortcuts.ts
index 00ee2df7a..9e80b8fc3 100644
--- a/ts/lib/shortcuts.ts
+++ b/ts/lib/shortcuts.ts
@@ -154,3 +154,42 @@ export function registerShortcut(
     document.addEventListener("keydown", handler);
     return (): void => document.removeEventListener("keydown", handler);
 }
+
+export interface ShortcutHandle {
+    onDestroy: () => void;
+    withShortcutSuffix: (string) => string;
+}
+
+/// Call within a component's onMount to assign the provided shortcut.
+export function useShortcut(
+    button: HTMLButtonElement,
+    shortcut = "",
+    optionalModifiers: Modifier[] | undefined = []
+): ShortcutHandle {
+    if (!shortcut) {
+        return {
+            onDestroy: () => {
+                return;
+            },
+            withShortcutSuffix: (s) => s,
+        };
+    }
+
+    const onDestroy = registerShortcut(
+        (event: KeyboardEvent) => {
+            button.dispatchEvent(new KeyboardEvent("click", event));
+            event.preventDefault();
+        },
+        shortcut,
+        optionalModifiers
+    );
+
+    const shortcutLabel = getPlatformString(shortcut);
+
+    return {
+        onDestroy,
+        withShortcutSuffix: (label: string | undefined) => {
+            return `${label ?? ""} (${shortcutLabel})`;
+        },
+    };
+}
```

Reply:
(or another possible alternative, if it made sense, would be to keep the existing WithShortcut, but use that in each component, instead of in the calling code?)

Reply:
I originally had each button take a `shortcutLabel`, which made for awkward functions like this:

```js
export function mergeTooltipAndShortcut(
    tooltip: string | undefined,
    shortcutLabel: string | undefined
): string | undefined {
    if (!tooltip && !shortcutLabel) {
        return undefined;
    }

    let buf = tooltip ?? "";
    if (shortcutLabel) {
        buf = `${buf} (${shortcutLabel})`;
    }
    return buf;
}
```

It's also not quite clear cut, how an item should deal with the shortcut label, e.g. `DropdownItem` uses it differently then buttons with tooltips. At some point I started using `appendInParentheses()`, you think that is sufficient?

Reply:
I'm coming from the Qt world where things like shortcuts, hiding and so on is all implemented on QWidget, and inherited by other components like push buttons, so at first glance a ShortcutIconButton feels a bit strange to me. But that may be trying to map what I'm used to with Qt into a different domain :-)

ShortcutIconButton keeps the implementation of IconButton cleaner, but makes things a bit less ergonomic if the majority of our IconButtons have shortcuts. I also wonder how this will combine with other features - for the sake of argument for example, imagine we decide we want to add a hiding ability as well. If the logic is inside IconButton with a separate prop, users could use the same widget with hiding on/off, and shortcuts on/off. If we expose separate components for this functionality, I fear it may become awkward if we need to have separate IconButton, ShortcutIconButton, HidableIconButton, and ShortcutAndHidableIconButtons. What do you think?

Reply:
I agree that is terrible:

>  IconButton, ShortcutIconButton, HidableIconButton, and ShortcutAndHidableIconButtons.

However that is also not a pretty option:

>  If the logic is inside IconButton with a separate prop, users could use the same widget with hiding on/off, and shortcuts on/off. 

Because it would lead to code that looks like this:

```javascript

{#if shortcuts && state}
    <WithShortcut>
        <WithState>
        <IconButton>
            ...
        </IconButton>
        </WithState>
    </WithShortcut>
{#else if shortcuts}
    <WithShortcut>
        <IconButton>
            ...
        </IconButton>
    </WithShortcut>
{#else if state}
    ...
{#else}
    ...
{/if}
```

Now, we could write the shortcut / state handling logic into each component directly, but then we'd have to duplicate the logic for each button, which I'd argue is the worst option to go with.

I've also intended to implement the hiding logic, which would make for 8 permutations altogether (if I remember my combinatorics correctly). That's make me think that the current implementation, with some helpers like `tooltip={appendInParentheses(tr.editingSetForegroundColor(), shortcutLabel)}` is not that bad after all. What do you think?

Reply:
We could avoid all the extra if statements if each "mixin" like WithShortcut just did a sensible thing when provided with a default value. useShortcut() above avoids adding a shortcut if an empty string is passed in, and a hideable mixin would behave as if it wasn't used if hidden=false - we'd just need to expose the prop in each component, and give it a default value.

> Now, we could write the shortcut / state handling logic into each component directly, but then we'd have to duplicate the logic for each button, which I'd argue is the worst option to go with.

But that's not really what the diff I posted above is doing, is it? The useShortcut() function behaves similarly to a mixin, with the logic in shortcuts.ts being reusable in other components as well. There are about 8 lines of boilerplate that each component would need to include, but that's mainly because Svelte doesn't give us an easier way to get a reference to the HTML element, and we can't have a helper function that calls onMount() itself. That's the appeal of using <WithShortcut ...> instead of useShortcut() - it would let us avoid some of that boilerplate.

Reply:
I think the crux of the issue is: should WithShortcut et al be placed inside each component, or should each component by wrapped by them in the calling code instead? I think the former option might be better, at least in cases like shortcuts and hiding, as:
- we'll be writing these components once, and using them many times in calling code. Having to wrap components in multiple mixins each time we use them is quite inconvenient and error-prone
- with default values for the props, each mixin can basically be a no-op if its functionality is not requested
- while it adds some boilerplate to each component implementation, the bulk of the logic is still contained in the mixin itself

Reply:
Looking at the current code, you can see that it includes yet another `mixin`, called `ButtonGroupItem`, which could not be included in the individual button themselves (as they have to be interacted with from `ButtonGroup`.

I didn't tackle this yet, but I have to say, I'd really dislike, making the mixins no-ops on demand. It would mean that we have to make the arguments optional, and we'd muddy the signature of the mixin. Also the idea of "we're invoking the Svelte component, but we don't want it to have any effect whatsoever" seems just wrong to me. I'll give the initial `useShortcut` suggestion a try.